crates/suins-indexer/src/indexer.rs

```diff
@@ -0,0 +1,219 @@
+    pub fn is_subdomain_wrapper(&self, object: &Object) -> bool {
+        object.struct_tag().is_some() && object.struct_tag().unwrap() == self.subdomain_wrapper_type
+    }
+
+    // Filter by owner.
+    // Owner has to be the TABLE of the registry.
+    // A table of that type can only have `Field<Domain,NameRecord> as a child so that check is enough to
+    // make sure we're dealing with a registry change.
+    pub fn is_name_record(&self, object: &Object) -> bool {
+        object.get_single_owner().is_some()
+            && object.get_single_owner() == Some(self.registry_table_id)
+    }
+
+    /// Parses the name record changes + subdomain wraps.
+    /// and pushes them into the supplied vector + hashmap.
+    ///
+    /// It is implemented in a way to do just a single iteration over the objects.
+    pub fn parse_name_record_changes(
+        &self,
+        objects: &Vec<&Object>,
+        name_record_changes: &mut Vec<NameRecordChange>,
+        sub_domain_wrappers: &mut HashMap<String, String>,
+    ) {
+        for &object in objects {
+            // Parse all the changes to a `NameRecord`
+            if self.is_name_record(object) {
+                let name_record: Field<Domain, NameRecord> = object
+                    .to_rust()
+                    .unwrap_or_else(|| panic!("Failed to parse name record for {:?}", object));
+
+                name_record_changes.push(NameRecordChange {
+                    field: name_record,
+                    field_id: object.id(),
+                });
+            }
+            // Parse subdomain wrappers and save them in our hashmap.
+            // Later, we'll save the id of the wrapper in the name record.
+            // NameRecords & their equivalent SubdomainWrappers are always created in the same PTB, so we can safely assume
+            // that the wrapper will be created on the same checkpoint as the name record and vice versa.
+            if self.is_subdomain_wrapper(object) {
+                let sub_domain: SubDomainRegistration = object.to_rust().unwrap();
+                sub_domain_wrappers.insert(
+                    sub_domain.nft.domain_name,
+                    sub_domain.id.id.bytes.to_string(),
+                );
+            };
+        }
+    }
+
+    /// For each input object, we're parsing the name record deletions
+    /// A deletion we want to track is a deleted object which is of `NameRecord` type.
+    /// Domain replacements do not count as deletions, but instead are an update to the latest state.
+    pub fn parse_name_record_deletions(
+        &self,
+        checkpoint: &CheckpointData,
+        removals: &mut Vec<String>,
+    ) {
+        // Gather all object ids that got deleted.
+        // This way, we can delete removed name records
+        // (detects burning of expired names or leaf names removal).
+        let deleted_objects: Vec<_> = checkpoint
```

Comment:
would a `Set` be a better container here?

Reply:
I don't think there can be a duplicate deletion for a single objectId?

Reply:
just pointing out we are doing a `contains` later and that seems to be the only way we use this?
in that sense felt we were doing more of a set operation than a vector

Reply:
Oh yeah. Great point. 😅